File: tests/test_binary_search_tree.py
from data_structures.binary_search_tree import BinarySearchTree


def build_tree(values):
    tree = BinarySearchTree()
    for v in values:
        tree.insert(v)
    return tree


def test_insert_and_search():
    tree = build_tree([5, 3, 8, 1, 4, 7, 9])
    for v in [5, 3, 8, 1, 4, 7, 9]:
        assert tree.search(v) is True
    assert tree.search(100) is False


def test_duplicate_insert_ignored():
    tree = build_tree([5, 3, 5, 3])
    assert len(tree) == 2


def test_in_order_is_sorted():
    values = [5, 3, 8, 1, 4, 7, 9, 2, 6]
    tree = build_tree(values)
    assert tree.in_order() == sorted(values)


def test_delete_leaf():
    tree = build_tree([5, 3, 8])
    tree.delete(3)
    assert tree.search(3) is False
    assert tree.in_order() == [5, 8]
    assert len(tree) == 2


def test_delete_node_with_one_child():
    tree = build_tree([5, 3, 8, 1])
    tree.delete(3)
    assert tree.search(3) is False
    assert tree.in_order() == [1, 5, 8]


def test_delete_node_with_two_children():
    tree = build_tree([5, 3, 8, 1, 4, 7, 9])
    tree.delete(3)
    assert tree.search(3) is False
    assert tree.in_order() == [1, 4, 5, 7, 8, 9]
    assert len(tree) == 6


def test_delete_root():
    tree = build_tree([5, 3, 8, 1, 4, 7, 9])
    tree.delete(5)
    assert tree.search(5) is False
    assert tree.in_order() == [1, 3, 4, 7, 8, 9]
    assert len(tree) == 6


def test_height_reflects_shape():
    empty = BinarySearchTree()
    assert empty.height() == -1

    single = build_tree([1])
    assert single.height() == 0

    # A sorted-order insertion degenerates into a linked list -- worth
    # asserting explicitly, since it's the whole reason balanced trees exist.
    degenerate = build_tree([1, 2, 3, 4, 5])
    assert degenerate.height() == 4

File: tests/test_binary_search_tree.cpp
#include "../include/data_structures/binary_search_tree.h"
#include "../include/test_framework.h"

TEST_CASE("bst: insert then in_order returns sorted values") {
    BinarySearchTree<int> tree;
    tree.insert(5);
    tree.insert(3);
    tree.insert(8);
    tree.insert(1);
    tree.insert(4);
    auto order = tree.in_order();
    std::vector<int> expected = {1, 3, 4, 5, 8};
    CHECK(order == expected);
}

TEST_CASE("bst: contains finds inserted values and rejects missing ones") {
    BinarySearchTree<int> tree;
    tree.insert(5);
    tree.insert(3);
    tree.insert(8);
    CHECK(tree.contains(5));
    CHECK(tree.contains(3));
    CHECK(!tree.contains(99));
}

TEST_CASE("bst: duplicate insert is a no-op (set semantics)") {
    BinarySearchTree<int> tree;
    tree.insert(5);
    tree.insert(5);
    CHECK(tree.size() == 1);
}

TEST_CASE("bst: remove a leaf node") {
    BinarySearchTree<int> tree;
    tree.insert(5);
    tree.insert(3);
    tree.insert(8);
    tree.remove(3);
    CHECK(!tree.contains(3));
    std::vector<int> expected = {5, 8};
    CHECK(tree.in_order() == expected);
}

TEST_CASE("bst: remove a node with two children preserves BST order") {
    BinarySearchTree<int> tree;
    for (int v : {5, 3, 8, 1, 4, 7, 9}) tree.insert(v);
    tree.remove(5);
    CHECK(!tree.contains(5));
    std::vector<int> expected = {1, 3, 4, 7, 8, 9};
    CHECK(tree.in_order() == expected);
}

TEST_CASE("bst: empty tree has height 0") {
    BinarySearchTree<int> tree;
    CHECK(tree.height() == 0);
    CHECK(tree.empty());
}

TEST_CASE("bst: height reflects shape -- sorted-order insertion degenerates to a line") {
    BinarySearchTree<int> degenerate;
    for (int v = 1; v <= 7; ++v) degenerate.insert(v);  // sorted order: degenerates
    CHECK(degenerate.height() == 7);  // O(n), the unbalanced worst case

    BinarySearchTree<int> balanced;
    for (int v : {4, 2, 6, 1, 3, 5, 7}) balanced.insert(v);  // level order: stays balanced
    CHECK(balanced.height() == 3);  // O(log n)
}
